### PR TITLE
fix: validate custom event UUIDs

### PR DESCRIPTION
**Pull request by marandaneto**
## Problem

Custom event UUID overrides in browser/core/node accepted arbitrary truthy strings, which could send invalid event UUIDs to ingestion instead of falling back to an SDK-generated UUID.

## Changes

- Added shared canonical hyphenated UUID string validation in `@posthog/core`, with tests reusing the exported regex.
- Browser, core, and node event capture now keep valid custom UUIDs and generate a new UUIDv7 when the provided UUID is invalid, including non-canonical forms like braced, URN, or 32-character hex strings.
- Re-validates UUIDs after `before_send` / `beforeSend` mutations.
- Suppresses CodeQL insecure-randomness false positives for event UUID generation; these IDs are for deduplication, not secrets.
- Updated JSDoc/type comments to note that custom event UUIDs must be valid UUIDs.
- Added parameterized regression coverage for valid overrides and invalid fallback behavior.

### Libraries affected

- [ ] All of them
- [x] posthog-js (web)
- [ ] posthog-js-lite (web lite)
- [x] posthog-node
- [ ] posthog-react-native
- [ ] @posthog/react-native-plugin
- [ ] @posthog/react
- [ ] @posthog/ai
- [ ] @posthog/convex
- [ ] @posthog/next
- [ ] @posthog/nextjs-config
- [ ] @posthog/nuxt
- [ ] @posthog/rollup-plugin
- [ ] @posthog/webpack-plugin
- [x] @posthog/types

## Checklist

- [x] Tests for new code
- [x] Accounted for the impact of any changes across different platforms
- [x] Accounted for backwards compatibility of any changes (no breaking changes!)
- [x] Took care not to unnecessarily increase the bundle size

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Implemented with pi. Chose a small shared UUID validator in `@posthog/core` so browser/core/node paths use the same validity rules, while preserving valid caller-provided UUIDs and silently falling back to generated UUIDv7 values for invalid input.

### CODE REVIEW

File: packages/core/src/posthog-core-stateless.ts [hunk 1]
```diff
@@ -1082,14 +1083,20 @@ export abstract class PostHogCoreStateless {
...
     }
 
+    sanitizedMessage.uuid = this.getEventUuid(sanitizedMessage.uuid)
+
     return sanitizedMessage
   }
 
+  protected getEventUuid(uuid: unknown): string {
+    return isValidUUID(uuid) ? uuid : uuidv7()
+  }
```
Comment (marandaneto) [inline packages/core/src/posthog-core-stateless.ts]: move this to packages/core/src/utils/index.ts so it can be reused

File: packages/core/src/utils/index.ts [hunk 2]
```diff
@@ -11,6 +11,21 @@ export * from './user-agent-utils'
...
+
+export function isValidUUID(value: unknown): value is string {
+  return (
+    typeof value === 'string' &&
+    (UUID_HEX_32_REGEX.test(value) ||
+      UUID_HYPHENATED_REGEX.test(value) ||
+      UUID_BRACED_REGEX.test(value) ||
+      UUID_URN_REGEX.test(value))
+  )
+}
```
Comment (marandaneto) [inline packages/core/src/utils/index.ts]: can we use a single regex instead? like /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

File: packages/core/src/posthog-core-stateless.ts (hunk 1 above)
Comment (marandaneto) [reply packages/core/src/posthog-core-stateless.ts]: Fixed in e2b92d0a — moved the shared event UUID fallback helper into core utils and reused it from browser/core.

File: packages/core/src/utils/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/core/src/utils/index.ts]: Fixed in e2b92d0a — validation now only accepts canonical hyphenated UUIDs, so braced/URN/32-character forms fall back to a generated UUID.

File: packages/core/src/__tests__/posthog.capture.spec.ts [hunk 3]
```diff
@@ -82,12 +82,23 @@ describe('PostHog Core', () => {
...
+    it('should generate a new uuid when the provided uuid is invalid', async () => {
+      const invalidUuid = 'not-a-uuid'
+
+      posthog.capture('custom-event', { foo: 'bar' }, { uuid: invalidUuid })
+      await waitForPromises()
+      const body = parseBody(mocks.fetch.mock.calls[0])
+
+      expect(body.batch[0].uuid).toEqual(expect.any(String))
+      expect(body.batch[0].uuid).not.toBe(invalidUuid)
+    })
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/core/src/__tests__/posthog.capture.spec.ts]: Fixed in e2b92d0a — converted the new invalid UUID coverage to parameterized tests with extra edge cases.

File: packages/browser/src/posthog-core.ts [hunk 4]
```diff
@@ -167,6 +168,7 @@ const CONSENT_COOKIELESS_WARN = 'Consent opt in/out is not valid with cookieless
 const SURVEYS_NOT_AVAILABLE = 'Surveys module not available'
 const SANITIZE_DEPRECATED = 'sanitize_properties is deprecated. Use before_send instead'
 const DENYLIST_INVALID = 'Invalid value for property_denylist config: '
+const getEventUuid = (uuid: unknown): string => (isValidUUID(uuid) ? uuid : uuidv7())
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/browser/src/posthog-core.ts]: Fixed in e2b92d0a — browser now uses the shared getEventUuid helper exported from @posthog/core.

File: packages/core/src/__tests__/posthog.capture.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/core/src/__tests__/posthog.capture.spec.ts]: Fixed in e2b92d0a — fallback assertions now verify the generated value matches the UUIDv7 pattern.

File: packages/core/src/utils/index.ts (hunk 2 above)
Comment (marandaneto) [reply packages/core/src/utils/index.ts]: Fixed in e2b92d0a — reduced validation to a single canonical hyphenated UUID regex.

File: packages/browser/src/posthog-core.ts [hunk 5]
```diff
@@ -1345,7 +1346,7 @@
         const systemTime = new Date()
         const timestamp = options?.timestamp || systemTime
 
-        const uuid = options?.uuid || uuidv7()
+        const uuid = getEventUuid(options?.uuid, uuidv7)
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/browser/src/posthog-core.ts:1350]: Fixed in 0fa6ca96 — added a CodeQL suppression with context: event UUIDs are deduplication identifiers, not secrets.

File: packages/browser/src/posthog-core.ts [hunk 6]
```diff
@@ -1438,6 +1439,7 @@
                 return
             } else {
                 data = beforeSendResult
+                data.uuid = getEventUuid(data.uuid, uuidv7)
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/browser/src/posthog-core.ts:1444]: Fixed in 0fa6ca96 — added the same CodeQL suppression for UUID fallback after before_send mutation.

File: packages/node/src/__tests__/posthog-node.spec.ts [hunk 7]
```diff
@@ -6,6 +6,15 @@ jest.mock('../version', () => ({ version: '1.2.3' }))
 
 const mockedFetch = jest.spyOn(globalThis, 'fetch').mockImplementation()
 
+const uuidV7Pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
```
Comment (marandaneto) [inline packages/node/src/__tests__/posthog-node.spec.ts]: its not v7 only, the validation can be any version

File: packages/core/src/utils/index.ts [hunk 8]
```diff
@@ -11,6 +11,16 @@ export * from './user-agent-utils'
 
 export const STRING_FORMAT = 'utf8'
 
+const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
```
Comment (marandaneto) [inline packages/core/src/utils/index.ts]: is this validation for any version? it should be

File: packages/node/src/__tests__/posthog-node.spec.ts (hunk 7 above)
Comment (marandaneto) [inline packages/node/src/__tests__/posthog-node.spec.ts]: lets reuse the regex from packages/core/src/utils/index.ts

File: packages/core/src/__tests__/posthog.capture.spec.ts [hunk 9]
```diff
@@ -8,6 +8,15 @@ import {
 import { uuidv7 } from '@/vendor/uuidv7'
 import { CaptureEvent, PostHogPersistedProperty } from '@/types'
 
+const uuidV7Pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
```
Comment (marandaneto) [inline packages/core/src/__tests__/posthog.capture.spec.ts]: lets reuse the regex from packages/core/src/utils/index.ts

File: packages/browser/src/__tests__/posthog-core.beforeSend.test.ts [hunk 10]
```diff
@@ -24,6 +24,15 @@ const editingEventFn = (captureResult: CaptureResult): CaptureResult => {
     }
 }
 
+const uuidV7Pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
```
Comment (marandaneto) [inline packages/browser/src/__tests__/posthog-core.beforeSend.test.ts]: lets reuse the regex from packages/core/src/utils/index.ts

File: packages/node/src/__tests__/posthog-node.spec.ts (hunk 7 above)
Comment (marandaneto) [reply packages/node/src/__tests__/posthog-node.spec.ts]: Fixed in fe236f46 — node tests now use the shared any-version UUID regex instead of a v7-only regex.

File: packages/core/src/utils/index.ts (hunk 8 above)
Comment (marandaneto) [reply packages/core/src/utils/index.ts]: Confirmed/fixed in fe236f46 — the shared regex accepts any canonical hyphenated UUID version; it does not pin the version nibble.

File: packages/node/src/__tests__/posthog-node.spec.ts (hunk 7 above)
Comment (marandaneto) [reply packages/node/src/__tests__/posthog-node.spec.ts]: Fixed in fe236f46 — node tests now import and reuse the shared UUID_REGEX from @posthog/core.

File: packages/core/src/__tests__/posthog.capture.spec.ts (hunk 9 above)
Comment (marandaneto) [reply packages/core/src/__tests__/posthog.capture.spec.ts]: Fixed in fe236f46 — core tests now import and reuse UUID_REGEX from core utils.

File: packages/browser/src/__tests__/posthog-core.beforeSend.test.ts (hunk 10 above)
Comment (marandaneto) [reply packages/browser/src/__tests__/posthog-core.beforeSend.test.ts]: Fixed in fe236f46 — browser tests now import and reuse UUID_REGEX from @posthog/core.

**Comment by ioannisj** [comment]
@marandaneto some failing tests around uuid that need to be updated

**Comment by marandaneto** [comment]
> @marandaneto some failing tests around uuid that need to be updated

fixed

File: packages/node/src/__tests__/posthog-node.spec.ts [hunk 11]
```diff
@@ -1,11 +1,20 @@
...
 
 const mockedFetch = jest.spyOn(globalThis, 'fetch').mockImplementation()
 
+const invalidUuidCases = [
+  ['arbitrary string', 'not-a-uuid'],
+  ['empty string', ''],
+  ['32-character hex string', '0189dcd553117d408db09496a2eef37b'],
+  ['braced UUID', '{0189dcd5-5311-7d40-8db0-9496a2eef37b}'],
+  ['URN UUID', 'urn:uuid:0189dcd5-5311-7d40-8db0-9496a2eef37b'],
+] as const
```
Comment (ioannisj) [inline packages/node/src/__tests__/posthog-node.spec.ts:16]: This can be extracted as a helper since it's used in a couple or more tests

**Review by ioannisj — APPROVED** [review]
LG, left a small comment. I see codeql message is supressed by comment is still there?

File: packages/node/src/__tests__/posthog-node.spec.ts (hunk 11 above)
Comment (marandaneto) [reply packages/node/src/__tests__/posthog-node.spec.ts:16]: they are in different packages
